components/engine_rocks_helper/src/sst_recovery.rs

```diff
@@ -0,0 +1,138 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::collections::BTreeMap;
+use std::collections::Bound::{Excluded, Included, Unbounded};
+use std::fmt::{Display, Formatter};
+use std::sync::{mpsc, Arc, Mutex};
+use std::time::{Duration, Instant};
+
+use engine_rocks::raw::*;
+use futures::executor::block_on;
+use futures::Future;
+use pd_client::PdClient;
+use protobuf::descriptorx::FieldDescriptorProtoExt;
+use raftstore::store::fsm::StoreMeta;
+use std::collections::HashMap;
+use tikv_util::{self, codec, set_panic_mark, thd_name, worker::*, *};
+
+const MAX: Duration = Duration::from_secs(30);
+
+pub fn create_sst_recovery_worker(
+    db: Arc<DB>,
+    pd_client: Arc<dyn PdClient>,
+    meta: Arc<Mutex<StoreMeta>>,
+) -> (FutureWorker<u64>, Scheduler<u64>) {
+    let worker = FutureWorker::new("sst-recovery");
+    worker
+        .start(RecoveryJob {
+            db,
+            meta,
+            pd_client,
+        })
+        .unwrap();
+    (worker, worker.scheduler())
+}
+
+// worker will run this job and try to receive any recovery task.
+pub struct RecoveryJob {
+    db: Arc<DB>,
+    meta: Arc<Mutex<StoreMeta>>,
+    pd_client: Arc<dyn PdClient>,
+    scheduling_regions: HashMap<u64, Instant>,
+}
+
+impl Runnable for RecoveryJob {
+    type Task = Task;
+
+    fn run(&mut self, sst: u64) {
+        self.generate_task(sst);
+
+        loop {
+            if self.scheduling_regions.is_empty() {
+                return;
+            }
+            for (region, time) in self.scheduling_regions.iter() {
+                if time < MAX {
+                    continue;
+                } else {
+                    self.check_scheduling_tasks();
+                }
+            }
+        }
+    }
+}
+
+impl RecoveryJob {
+    fn generate_task(&mut self, sst: u64) {
+        self.db.resume().unwrap();
```

Comment:
Repeated errors will occur again I guess, but eventually they will all converge to `scheduling_tasks` and check. Maybe I need add a global timer to the entire recovery job to prevent repeated errors from happening, so that tikv will never panic